### PR TITLE
Static IP configuration support

### PR DESCRIPTION
**Pull request by avtolstoy**
## TODO

1. Some automated tests
2. USB/BLE control request updates

### Description

See SDD: https://docs.google.com/document/d/1WjEtv3wgbmD6C93mm2qny-MG50iB2rR7hk8GNb6cuA4/edit#

### Steps to Test

See example apps below

#### WiFi

```c++
void loop() {
    if (Serial.available()) {
        char c = Serial.read();
        if (c == 'r') {
            auto conf = WiFi.getConfig("TestNet");
            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            for (const auto& addr: conf.dns()) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
        } else if (c == 'w') {
            auto conf = NetworkInterfaceConfig()
                .source(NetworkInterfaceConfigSource::STATIC, AF_INET)
                .address(IPAddress(192,168,1,222), IPAddress(255,255,255,0))
                .gateway(IPAddress(192,168,1,99))
                .dns(IPAddress(192,168,1,99))
                .dns(IPAddress(8,8,8,8))
                .profile("TestNet");

            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            auto dns = conf.dns();
            (void)dns;
            for (const auto& addr: dns) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
            WiFi.setConfig(conf);
        } else if (c == 'c') {
            Particle.connect();
        } else if (c == 'd') {
            Particle.disconnect();
            Network.disconnect();
        } else if (c == 'u') {
            unlink("/sys/network.dat");
        }
    }
}
```

#### Ethernet
```c++
void loop() {
    if (Serial.available()) {
        char c = Serial.read();
        if (c == 'r') {
            auto conf = Ethernet.getConfig();
            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            for (const auto& addr: conf.dns()) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
        } else if (c == 'w') {
            auto conf = NetworkInterfaceConfig()
                .source(NetworkInterfaceConfigSource::STATIC, AF_INET)
                .address(IPAddress(192,168,1,222), IPAddress(255,255,255,0))
                .gateway(IPAddress(192,168,1,99))
                .dns(IPAddress(192,168,1,99))
                .dns(IPAddress(8,8,8,8));

            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            auto dns = conf.dns();
            (void)dns;
            for (const auto& addr: dns) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
            Ethernet.setConfig(conf);
        } else if (c == 'c') {
            Particle.connect();
        } else if (c == 'd') {
            Particle.disconnect();
            Network.disconnect();
        } else if (c == 'u') {
            unlink("/sys/network.dat");
        }
    }
}
```

### References

- https://docs.google.com/document/d/1WjEtv3wgbmD6C93mm2qny-MG50iB2rR7hk8GNb6cuA4/edit#

---

### Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

**Comment by scott-brust** [comment]
Apologies for the stupid questions, but I want to make sure I am testing things correctly. 

I have a router with wifi + ethernet. I have 1) disabled DHCP on the router and 2) setup a static IP reservation for the ethernet MAC of my test Photon2 device.
Ive updated the example code with the reserved IP (`192.168.0.222`) and the default gateway/DNS for my router (a TP Link router) `192.168.0.1`. 
I stored this config (ie `w` command in app)
```
0000004152 [app] INFO: Addr: 192.168.0.222/24 (255.255.255.0) - family = 2
0000004179 [app] INFO: DNS: 192.168.0.1
0000004194 [app] INFO: DNS: 8.8.8.8
0000004207 [app] INFO: Gateway: 192.168.0.1
0000004223 [app] INFO: Source: 2/0
```

Then connected. Cloud comms looked good:
```
0000011381 [system] INFO: All handshake messages have been processed
```

I read back my Ethernet IP and can see its using the reserved IP:
```
0000017905 [app] INFO: MAC: 94:94:4A:04:08:EB
0000017922 [app] INFO: localIP: 192.168.0.222
0000017939 [app] INFO: DHCP: 0.0.0.0
0000017953 [app] INFO: Gateway IP: 192.168.0.1
0000017973 [app] INFO: Subnet Mask: 255.255.255.0
```
So far so good. Ill try wifi next

File: system/src/system_network_manager.cpp [hunk 1]
```diff
@@ -1003,6 +1048,201 @@ int NetworkManager::waitInterfaceOff(if_t iface, system_tick_t timeout) const {
...
+                LOG(ERROR, "6");
+                return false;
+            }
+
+            if (!pb_encode_tag_for_field(strm, field)) {
+                LOG(ERROR, "7");
+                return false;
+            }
+            if (!pb_encode_submessage(strm, PB_CTRL(Interface_fields), &pbInterface)) {
+                LOG(ERROR, "8");
```
Comment (technobly) [inline system/src/system_network_manager.cpp:1236]: Reminder to remove debugging LOG's.

File: wiring/inc/spark_wiring_posix_common.h [hunk 2]
```diff
@@ -73,8 +78,58 @@ inline void ipAddressPortToSockaddr(const IPAddress& addr, uint16_t port, struct
 #endif // HAL_IPv6
 }
 
+inline uint8_t netmaskToPrefixLength(struct sockaddr* saddr) {
+    if (saddr->sa_family == AF_INET) {
```
Comment (technobly) [inline wiring/inc/spark_wiring_posix_common.h:82]: Add `CHECK_TRUE(saddr, SYSTEM_ERROR_INVALID_ARGUMENT);`

**Review by technobly — CHANGES_REQUESTED** [review]
I tested ethernet shield with GPIO config on D2,D0,D1 with STATIC and DHCP configs.

File: wiring/inc/spark_wiring_posix_common.h (hunk 2 above)
Comment (avtolstoy) [reply wiring/inc/spark_wiring_posix_common.h:82]: Functions here are helpers so the requirement is that saddr is valid.

File: hal/network/ncp_client/realtek/rtl_ncp_client.cpp [hunk 3]
```diff
@@ -100,9 +100,8 @@ void wifi_set_country_code(void) {
             country_code_sdk = RTW_COUNTRY_AU;
             break;
         case WLAN_CC_WORLD: {
-            channel_plan = 0x20;
-            country_code_sdk = RTW_COUNTRY_WORLD;
-            break;
+            // No specific channel plan/country code for both 2.4/5GHz
+            return;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/ncp_client/realtek/rtl_ncp_client.cpp:104]: So, the gist of the problem is that there is no country code/channel plan in the SDK that enables everything for both 2.4GHz and 5GHz bands. I've tested and not setting anything at all actually makes it behave as expected i.e. allowing to use any channels within both bands.

File: hal/network/lwip/realtek/rtlncpnetif.cpp [hunk 4]
```diff
@@ -254,6 +254,18 @@ int RealtekNcpNetif::getPowerState(if_power_state_t* state) const {
...
+int RealtekNcpNetif::getCurrentProfile(spark::Vector<char>* profile) const {
+    const auto client = wifiMan_->ncpClient();
+    CHECK_TRUE(client, SYSTEM_ERROR_UNKNOWN);
+
+    CHECK_TRUE(client->connectionState() == NcpConnectionState::CONNECTED, SYSTEM_ERROR_INVALID_STATE);
+
+    WifiNetworkInfo info;
+    CHECK(client->getNetworkInfo(&info));
+    *profile = spark::Vector<char>(info.ssid(), strlen(info.ssid()));
+    return 0;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/lwip/realtek/rtlncpnetif.cpp:266]: Let's all agree once again that it's not a requirement to `SYSTEM_ERROR_NONE` and shorter `0` is a valid value meaning the same thing.

File: hal/network/lwip/esp32/esp32ncpnetif.cpp [hunk 5]
```diff
@@ -224,6 +224,18 @@ void Esp32NcpNetif::loop(void* arg) {
...
+int Esp32NcpNetif::getCurrentProfile(spark::Vector<char>* profile) const {
+    const auto client = wifiMan_->ncpClient();
+    CHECK_TRUE(client, SYSTEM_ERROR_UNKNOWN);
+
+    CHECK_TRUE(client->connectionState() == NcpConnectionState::CONNECTED, SYSTEM_ERROR_INVALID_STATE);
+
+    WifiNetworkInfo info;
+    CHECK(client->getNetworkInfo(&info));
+    *profile = spark::Vector<char>(info.ssid(), strlen(info.ssid()));
+    return 0;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/lwip/esp32/esp32ncpnetif.cpp:236]: Let's all agree once again that it's not a requirement to `SYSTEM_ERROR_NONE` and shorter `0` is a valid value meaning the same thing.

File: hal/network/ncp_client/realtek/rtl_ncp_client.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/ncp_client/realtek/rtl_ncp_client.cpp:104]: Just to clarify: the default is still US, if one wants to set to `WLAN_CC_WORLD` (any channels) right now this doesn't work for 5GHz band, `RTW_COUNTRY_WORLD` and channel plan = 0x20 (also world) only enables all channels within 2.4GHz band.

File: hal/network/lwip/esp32/esp32ncpnetif.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (technobly) [reply hal/network/lwip/esp32/esp32ncpnetif.cpp:236]: Less typing? Ok :D

[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/lwip/esp32/esp32ncpnetif.cpp:236]: Yup, just to type less. It's a common consensus in APIs that `0` == ok and we've being doing either 0 or SYSTEM_ERROR_NONE everywhere, so not a new thing either.

File: hal/network/ncp_client/realtek/rtl_ncp_client.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (technobly) [reply hal/network/ncp_client/realtek/rtl_ncp_client.cpp:104]: So with your change, if WLAN_CC_WORLD is set, all channels are enabled for 2.4 & 5Ghz?  If so could you add that in the comments there?

[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/ncp_client/realtek/rtl_ncp_client.cpp:104]: Yes, that's correct. I'll add a bit more of an explanation.

**Comment by avtolstoy** [comment]
Outstanding review comments resolved. PR also updated with:
1. DNS priority list fixes
2. Support for DHCP gateway/dns list override

Quick and dirty app with some additional examples

```c++
/*
 * Copyright (c) 2019 Particle Industries, Inc.  All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see <http://www.gnu.org/licenses/>.
 */

#include "application.h"
#include "resolvapi.h"

SYSTEM_MODE(SEMI_AUTOMATIC);
SYSTEM_THREAD(ENABLED);

SerialLogHandler dbg(LOG_LEVEL_ALL);

/* executes once at startup */
void setup() {
    wlan_set_country_code(WLAN_CC_WORLD, nullptr);
}

/* executes continuously after setup() runs */
void loop() {
    if (Serial.available()) {
        char c = Serial.read();
        if (c == 'r') {
            auto conf = WiFi.getConfig("TestNetwork");
            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            for (const auto& addr: conf.dns()) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
        } else if (c == 'w') {
            // Full static configuration
            auto conf = NetworkInterfaceConfig()
                .source(NetworkInterfaceConfigSource::STATIC, AF_INET)
                .address(IPAddress(192,168,1,222), IPAddress(255,255,255,0))
                .gateway(IPAddress(192,168,1,99))
                .dns(IPAddress(192,168,1,99))
                .dns(IPAddress(1,1,1,1))
                .profile("TestNetwork");

            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            auto dns = conf.dns();
            (void)dns;
            for (const auto& addr: dns) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
            Log.info("setConfig = %d", WiFi.setConfig(conf));
        } else if (c == 'h') {
            // DHCP + DNS and gateway override
            auto conf = NetworkInterfaceConfig()
                .source(NetworkInterfaceConfigSource::DHCP, AF_INET)
                //.address(IPAddress(192,168,1,222), IPAddress(255,255,255,0))
                .gateway(IPAddress(192,168,1,99))
                .dns(IPAddress(8,8,8,8))
                .dns(IPAddress(1,1,1,1))
                .profile("TestNetwork");

            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            auto dns = conf.dns();
            (void)dns;
            for (const auto& addr: dns) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
            Log.info("setConfig = %d", WiFi.setConfig(conf));
        } else if (c == 'e') {
            // DHCP + DNS override and NO gateway (local only comms)
            auto conf = NetworkInterfaceConfig()
                .source(NetworkInterfaceConfigSource::DHCP, AF_INET)
                //.address(IPAddress(192,168,1,222), IPAddress(255,255,255,0))
                .gateway(IPAddress(0,0,0,0))
                .dns(IPAddress(8,8,8,8))
                .dns(IPAddress(1,1,1,1))
                .profile("TestNetwork");

            for (const auto& addr: conf.addresses()) {
                Log.info("Addr: %s/%u (%s) - family = %d", addr.address().address().toString().c_str(), (unsigned)addr.prefixLength(), addr.mask().address().toString().c_str(), addr.family());
            }
            auto dns = conf.dns();
            (void)dns;
            for (const auto& addr: dns) {
                Log.info("DNS: %s", addr.toString().c_str());
            }
            Log.info("Gateway: %s", conf.gateway().toString().c_str());
            Log.info("Source: %d/%d", (int)conf.source(AF_INET), (int)conf.source(AF_INET6));
            Log.info("setConfig = %d", WiFi.setConfig(conf));
        } else if (c == 'n') {
            // Dumps DNS list in priority order
            resolv_dns_servers* dns = {};
            resolv_get_dns_servers(&dns);
            // Take only the first one
            for (auto s = dns; s != nullptr; s = s->next) {
                // Only IPv4 for now
                SockAddr a(s->server);
                Log.info("DNS: %s", a.toString().c_str());
            }
            resolv_free_dns_servers(dns);
        } else if (c == 'c') {
            Particle.connect();
        } else if (c == 'd') {
            Particle.disconnect();
            Network.disconnect();
        } else if (c == 'u') {
            unlink("/sys/network.dat");
        } else if (c == 'a') {
            // Dumps current interface addresses
            if_t iface = nullptr;
            if_get_by_index(WiFi, &iface);
            if_addrs* ifAddrList = nullptr;
            if_get_addrs(iface, &ifAddrList);
            for (if_addrs* i = ifAddrList; i; i = i->next) {
                if (!i->if_addr) {
                    continue;
                }
                SockAddr addr(i->if_addr->addr);
                SockAddr mask(i->if_addr->netmask);
                SockAddr gw(i->if_addr->gw);
                auto family = i->if_addr->addr ? i->if_addr->addr->sa_family : -1;
                Log.info("Addr: family=%d %s/%u (%s) gw=%s", family, addr.toString().c_str(), i->if_addr->prefixlen, mask.toString().c_str(), gw.toString().c_str());
            }
            if_free_if_addrs(ifAddrList);
        }
    }
}
```

## TODO

Wrap-up automated tests.

**Review by technobly — APPROVED** [review]
Test apps and integrations tests passing for me, and while Ethernet GPIO is reconfigured for D2, D0, D1.

**Pull request merged**